src/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/HashAlgorithm.cs

```diff
@@ -110,6 +112,35 @@ public byte[] ComputeHash(Stream inputStream)
             return CaptureHashCodeAndReinitialize();
         }
 
+        public Task<byte[]> ComputeHashAsync(
+            Stream inputStream,
+            CancellationToken cancellationToken = default)
+        {
+            if (inputStream == null)
+                throw new ArgumentNullException(nameof(inputStream));
+            if (_disposed)
+                throw new ObjectDisposedException(null);
+
+            return ComputeHashAsyncCore(inputStream, cancellationToken);
+        }
+
+        private async Task<byte[]> ComputeHashAsyncCore(
+            Stream inputStream,
+            CancellationToken cancellationToken)
+        {
+            // Use ArrayPool.Shared instead of CryptoPool because the array is passed out.
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
+
+            int bytesRead;
+            while ((bytesRead = await inputStream.ReadAsync(buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false)) > 0)
+            {
+                HashCore(buffer, 0, bytesRead);
+            }
+
+            ArrayPool<byte>.Shared.Return(buffer, clearArray: true);
```

Comment:
Should we maintain a high water mark and only clear that much?

Reply:
Sure.  I've gone ahead and changed ComputeHash(Stream) to do the same, so they match.